// express-e2e/lib/utils/cookie.js
'use strict';

/***********************************
 *** START ~ Module dependencies ***
 ***********************************/

// Global dependencies


// Project dependencies
const dateUtil = require('./date');


/***********************************
 *** END ~ Module dependencies ***
 ***********************************/

const createCookie = (res, key, val, opt) => {
    opt = opt || {};
    res.cookie(key, val, opt);
};

exports.parse = (cookieString) => {
    let cookieJson = {};
    let cookies = cookieString.split(';');

    for(let cookie of cookies) {
        let kvp = cookie.trim().split('=');
        cookieJson[kvp[0]] = kvp[1];
    }

    return cookieJson;
};

exports.createInsecureCookie = (res, key, val) => {
    createCookie(res, key, val);
};

exports.createCookie = (req, res, key, val) => {
    let config = require('../../index').config();
    let expiry = new Date();

    dateUtil.addSeconds(expiry, config.jwt.expiresIn);

    let opt = {
        expires: expiry,
        httpOnly: true
    };

    if (req.isSecure) {
        opt.secure = true;
    }

    createCookie(res, key, val, opt);
};

exports.createSessionCookie = (req, res, key, val) => {
    let opt = {
        httpOnly: true
    };

    if (req.isSecure) {
        opt.secure = true;
    }

    createCookie(res, key, val, opt);
};

exports.createExpiredCookie = (req, res, key, val) => {
    let expiry = new Date();
    dateUtil.addYear(expiry, -10);

    let opt = {
        expires: expiry,
        maxAge: -365000,
        httpOnly: true
    };

    if (req.isSecure) {
        opt.secure = true;
    }

    createCookie(res, key, val, opt);
};
